scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ml.dmlc.mxnet.infer
+
+import java.util.concurrent._
+
+import org.slf4j.LoggerFactory
+
+private[infer] trait MXNetHandler {
+
+  def execute[T](f: => T): T
+
+  val executor: ExecutorService
+
+}
+
+private[infer] object MXNetHandlerType extends Enumeration {
+
+  type MXNetHandlerType = Value
+  val SingleThreadHandler = Value("MXNetSingleThreadHandler")
+  val OneThreadPerModelHandler = Value("MXNetOneThreadPerModelHandler")
+}
+
+private[infer] class MXNetThreadPoolHandler(numThreads: Option[Int] = Some(1))
+  extends MXNetHandler {
+  private val logger = LoggerFactory.getLogger(classOf[MXNetThreadPoolHandler])
+
+  private val threadFactory = new ThreadFactory {
+
+    override def newThread(r: Runnable): Thread = new Thread(r) {
```

Comment:
This will create all threads with the same name, which will make them difficult (or in some contexts impossible) to tell apart. If you're going to continue to support multiple threads consider incrementing a counter whenever you vend a thread and appending the number to the name (this is the behaviour of Guava's [ThreadFactoryBuilder](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/ThreadFactoryBuilder.html) for example).